luavm/src/lib/lua_values/mod.rs

```diff
@@ -0,0 +1,754 @@
+mod lua_table;
+mod tagging;
+
+use self::{lua_table::LuaTable, tagging::*};
+use errors::LuaError;
+use gc::{Finalize, Gc, Trace};
+
+/// Represents a value in Lua.
+#[derive(Debug)]
+pub struct LuaVal {
+    encoded_ptr: usize,
+    /// whether this should drop the pointer or not
+    owns: bool,
+}
+
+impl Finalize for LuaVal {}
+unsafe impl Trace for LuaVal {
+    custom_trace!(this, {
+        // only tables have a Gc inside, so there is no need to mark anything else
+        if let LuaValKind::TABLE = this.kind() {
+            mark(unsafe { &*table_ptr(this.encoded_ptr) });
+        }
+    });
+}
+
+impl LuaVal {
+    /// Create an empty LuaVal which is equivalent to Nil.
+    pub fn new() -> LuaVal {
+        LuaVal {
+            encoded_ptr: 0,
+            owns: true,
+        }
+    }
+
+    /// Copies the pointer into a new LuaVal.
+    fn shallow_copy(&self) -> LuaVal {
+        LuaVal {
+            encoded_ptr: self.encoded_ptr,
+            owns: false,
+        }
+    }
+
+    /// Returns the type of the value store in the pointer.
+    fn kind(&self) -> LuaValKind {
+        LuaValKind::from(self.encoded_ptr)
+    }
+
+    /// Returns true if this value can be considered a float, or false otherwise.
+    pub fn is_float(&self) -> bool {
+        match self.kind() {
+            LuaValKind::FLOAT | LuaValKind::STR => true,
+            _ => false,
+        }
+    }
+
+    /// Attempts to convert this value to a float.
+    pub fn to_float(&self) -> Result<f64, LuaError> {
+        unsafe {
+            match self.kind() {
+                LuaValKind::INT => Ok(*(i64_ptr(self.encoded_ptr)) as f64),
+                LuaValKind::FLOAT => Ok(*(f64_ptr(self.encoded_ptr))),
+                LuaValKind::STR => (*(str_ptr(self.encoded_ptr)))
+                    .parse()
+                    .map_err(|_| LuaError::FloatConversionErr),
+                _ => Err(LuaError::FloatConversionErr),
+            }
+        }
+    }
+
+    /// Attempts to convert this value to an integer.
+    pub fn to_int(&self) -> Result<i64, LuaError> {
+        unsafe {
+            match self.kind() {
+                LuaValKind::INT => Ok(*(i64_ptr(self.encoded_ptr))),
+                _ => Err(LuaError::IntConversionErr),
+            }
+        }
+    }
+
+    /// Sets this to a new value.
+    pub fn set(&mut self, mut val: LuaVal) {
+        // exchange the pointers so that when `val` goes out of scope the old value
+        // is released and not leaked
+        let new_address = val.encoded_ptr;
```

Comment:
`new_address` is a clunky name and probably not an accurate one either. Can we use something shorter?

Reply:
OK.